### PR TITLE
Interhand 3d dataset

### PR DESCRIPTION
**Pull request by liuxin9608**

### CODE REVIEW

File: mmpose/core/evaluation/top_down_eval.py [hunk 1]
```diff
@@ -245,8 +245,7 @@ def keypoint_epe(pred, gt, mask):
 
     distances = _calc_distances(
         pred, gt, mask,
-        np.tile(
-            np.array([[1 for i in range(pred.shape[2])]]), (pred.shape[0], 1)))
+        np.tile(np.array([[1] * pred.shape[2]]), (pred.shape[0], 1)))
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:248]: Is this an all-one matrix?

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+    """
+
+    def __init__(self,
+                 ann_file,
+                 camera_file,
+                 joint_file,
+                 img_prefix,
+                 data_cfg,
+                 pipeline,
+                 use_gt_root_depth=True,
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:86]: This arg is missing in the docstring

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+
+        Note:
+            batch_size: N
+            num_keypoints: K
+            heatmap height: H
+            heatmap width: W
+
+        Args:
+            outputs (list(preds, boxes, image_path, output_heatmap))
+                outputs (list(preds, boxes, image_path, output_heatmap))
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:306]: duplicated line?

Comment (innerlee) [reply mmpose/datasets/datasets/hand/interhand3d_dataset.py:306]: btw it seems that output is a dict

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+        allowed_metrics = ['MRRPE', 'MPJPE', 'Handedness_acc']
+        for metric in metrics:
+            if metric not in allowed_metrics:
+                raise KeyError(f'metric {metric} is not supported')
+
+        res_file = os.path.join(res_folder, 'result_keypoints.json')
+
+        kpts = []
+        for output in outputs:
+            if 'preds' in output:
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:332]: assume that output is a dict. you can write
```python
preds = output.get('preds')
```

**Comment by innerlee** [comment]
kindly ping @liuxin9608

File: mmpose/core/evaluation/top_down_eval.py [hunk 5]
```diff
@@ -242,8 +242,10 @@ def keypoint_epe(pred, gt, mask):
     Returns:
         float: Average end-point error.
     """
+
     distances = _calc_distances(
-        pred, gt, mask, np.tile(np.array([[1, 1]]), (pred.shape[0], 1)))
+        pred, gt, mask,
+        np.ones((pred.shape[0], pred.shape[2]), dtype=np.float32))
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:248]: So `pred.shape[2] == 2`?

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py (hunk 2 above)
Comment (innerlee) [reply mmpose/datasets/datasets/hand/interhand3d_dataset.py:89]: joint_file etc. are still mmissing

File: mmpose/core/evaluation/top_down_eval.py (hunk 5 above)
Comment (liuxin9608) [reply mmpose/core/evaluation/top_down_eval.py:248]: No.

Comment (liuxin9608) [reply mmpose/core/evaluation/top_down_eval.py:248]: 2 or 3

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+            world_coord (ndarray[3, N]): 3D joints coordinates
+                in the world coordinate system
+            R (ndarray[3, 3]): camera rotation matrix
+            T (ndarray[3]): camera position (x, y, z)
+
+        Returns:
+            cam_coord (ndarray[3, N]): 3D joints coordinates
+                in the camera coordinate system
+        """
+        cam_coord = np.dot(R, world_coord - T)
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:157]: > `world_coord - T`

 a [3, N] array minus a [3] array is likely to be an error

Comment (liuxin9608) [reply mmpose/datasets/datasets/hand/interhand3d_dataset.py:157]: T is shaped [3, 1]

File: mmpose/core/evaluation/top_down_eval.py (hunk 5 above)
Comment (innerlee) [reply mmpose/core/evaluation/top_down_eval.py:248]: Okay docstring is not exact

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+    @staticmethod
+    def _process_handtype(hand_type):
+        if hand_type == 'right':
+            return np.array([1, 0], dtype=int)
+        elif hand_type == 'left':
+            return np.array([0, 1], dtype=int)
+        elif hand_type == 'interacting':
+            return np.array([1, 1], dtype=int)
+        else:
+            assert 0, print('Not support hand type: ' + hand_type)
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:193]: print after assertion?

Comment (innerlee) [reply mmpose/datasets/datasets/hand/interhand3d_dataset.py:193]: f-string btw

File: mmpose/core/evaluation/top_down_eval.py (hunk 5 above)
Comment (liuxin9608) [reply mmpose/core/evaluation/top_down_eval.py:249]: fixed.

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+            ann_id = self.coco.getAnnIds(imgIds=img_id, iscrowd=False)
+            ann = self.coco.loadAnns(ann_id)[0]
+            img = self.coco.loadImgs(img_id)[0]
+
+            capture_id = str(img['capture'])
+            camera_name = img['camera']
+            frame_idx = str(img['frame_idx'])
+            image_file = os.path.join(self.img_prefix, self.id2name[img_id])
+
+            camera_pos, camera_rot = np.array(
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:229]: split camera_pos and camera_rot into two lines

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+                cameras[capture_id]['campos'][camera_name],
+                dtype=np.float32), np.array(
+                    cameras[capture_id]['camrot'][camera_name],
+                    dtype=np.float32)
+            focal, principal_pt = np.array(
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:234]: and here

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+                    dtype=np.float32)
+            joint_world = np.array(
+                joints[capture_id][frame_idx]['world_coord'], dtype=np.float32)
+            joint_cam = self._world2cam(
+                joint_world.transpose(1, 0), camera_rot,
+                camera_pos.reshape(3, 1)).transpose(1, 0)
+            joint_img = self._cam2pixel(joint_cam, focal, principal_pt)[:, :2]
+
+            joint_valid = np.array(
+                ann['joint_valid'], dtype=np.float32).reshape(-1)
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:247]: `reshape(-1)` -> `flatten()`?

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 11]
```diff
@@ -0,0 +1,534 @@
...
+                in the camera coordinate system
+        """
+        x = (pixel_coord[:, 0] - c[0]) / f[0] * pixel_coord[:, 2]
+        y = (pixel_coord[:, 1] - c[1]) / f[1] * pixel_coord[:, 2]
+        z = pixel_coord[:, 2]
+        cam_coord = np.concatenate((x[:, None], y[:, None], z[:, None]), 1)
+        return cam_coord
+
+    @staticmethod
+    def _process_handtype(hand_type):
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:185]: This "process" looks like an encoding.
If it happens to be called encoding in paper, then we can rename it to `_encode_handtype`

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+            joint_img = self._cam2pixel(joint_cam, focal, principal_pt)[:, :2]
+
+            joint_valid = np.array(
+                ann['joint_valid'], dtype=np.float32).reshape(-1)
+            hand_type = self._process_handtype(ann['hand_type'])
+            hand_type_valid = ann['hand_type_valid']
+
+            if self.use_gt_root_depth:
+                bbox = np.array(ann['bbox'], dtype=np.float32)
+                center, scale = self._xywh2cs(*bbox, 1.25)
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:253]: add a note for the number 1.25

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py [hunk 13]
```diff
@@ -0,0 +1,534 @@
...
+                        np.zeros(self.ann_info['num_joints'], dtype=bool))
+                    all_masks.append(mask)
+
+            if 'Handedness_acc' in metrics:
+                pred_hand_type = np.array(pred['hand_type'], dtype=int)
+                preds_hand_type.append(pred_hand_type)
+                gts_hand_type.append(item['hand_type'])
+                hand_type_masks.append(item['hand_type_valid'] > 0)
+
+        gts_rel_root = np.array(gts_rel_root)
```
Comment (innerlee) [inline mmpose/datasets/datasets/hand/interhand3d_dataset.py:500]: Make the datatype explict

**Comment by innerlee** [comment]
Some minor comments.

How about the training accuracy? When will the ckpts be ready?

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py (hunk 13 above)
Comment (innerlee) [reply mmpose/datasets/datasets/hand/interhand3d_dataset.py:500]: E.g., there are some variables with `mask` in their names. Then one may wonder their types: bool? int? float32?

**Comment by liuxin9608** [comment]
> Some minor comments.
> 
> How about the training accuracy? When will the ckpts be ready?

The training has not started yet. There are also many PRs to commit.

**Comment by innerlee** [comment]
I see. If more prs are needed for training, then we can simply merge this, and block on the last pr.

File: mmpose/datasets/datasets/hand/interhand3d_dataset.py (hunk 13 above)
Comment (innerlee) [reply mmpose/datasets/datasets/hand/interhand3d_dataset.py]: The type can be changed if the append operation is not careful. e.g.

image

**Review by innerlee — APPROVED** [review]

**Pull request merged**